### ISSUE TITLE
Websockets Authorization

### ISSUE THREAD

**Issue by manast** [uNetworking/uWebSockets.js#160]
Is there any best practice for implementing websockets authorization with uWS? 
In previous versions of the library it was possible to check a bearer token in the req.headers of the initial HTTP request before upgrading to websockets, then respond with status 401 oer whatever. Is this still possible somehow? currently I am doing the authorization in the ws callback like this:
```ts
app
  .ws("/*", {
    /* Options */
    compression: CompressOptions.SHARED_COMPRESSOR,
    maxPayloadLength: 16 * 1024 * 1024,
    idleTimeout: 10,
    open: (ws, req) => {
      console.log("A WebSocket connected via URL: " + req.getUrl() + "!");

      console.log(ws, req.getHeader("authorization"));

      const authorization = req.getHeader("authorization");

      if (authorization) {
        const token = authorization.split(" ")[1];
        if (token) {
          ws.userId = getUserIdFromToken;
          if(ws.userId) {
             return;
        }
      }
      // 404 is not a valid websocket error code
      ws.end(404, "Unauthorized");
    },
```
Problem is that there are not suitable standard websocket error codes that express missing authorization or similar. Also it is not as clean as just doing the authorization before even starting the ws connection.

**Comment by alexhultman** [comment]
Upgrade is simplified to just give you "open" event. You can close immediately or send a message then close. Is that a problem?

**Comment by manast** [comment]
It is not a problem per se, it can be done using the workaround that you described which is more or less what the code above does. However it seems more elegant to be able to return standard HTTP codes before the upgrade to websockets take place. Also, it forces to use a non-standard code for representing lack of authorization, or just return code 1000 and send a proper message: 
https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent

**Comment by sosdolar** [comment]
Similar if not the same point of issue #108

**Comment by manast** [comment]
Another thing to consider with the proposed workaround: the open event on the client is emitted before the open event in the server had any chance to close due to missing authorization.

**Comment by ibc** [comment]
WebSocket spec does not define any auth mechanism. It's up to the app to use Cookies (that are possible in the HTTP handshake request) or via custom WS URI query params. The server may check those before it decides to accept a connection and send a 403 if it rejects it.

Note: the browser will not get access to the specific HTTP 403 status code but a generic "error" event. That's how the W3 WebSocket API is defined.

**Comment by alexhultman** [comment]
> Another thing to consider with the proposed workaround: the open event on the client is emitted before the open event in the server had any chance to close due to missing authorization.

This is not true. The server's open event is emitted before the client's. Force closing the connection ws->close() will hinder the upgrade from reaching the client.

The only thing you cannot do, is sending a custom http response.

**Comment by ibc** [comment]
Can the server perform an async action (DB query for instance) that takes some seconds before calling ws.close() so it will prevent HTTP 101?

**Comment by alexhultman** [comment]
No. And any "fix" to this would be limited to "non-async" actions.

**Comment by ibc** [comment]
Is this fixed behavior? There are tons of use cases in which it's requiree to perform an async action to decide whether a WS connection must be established or not. And not, establishing the connection and later closing it (after some WS messages) is not an option in many of those cases.

**Comment by ibc** [comment]
By reading comments above it seems that those use cases are not known, so let me explain one:

```
Client     Nginx      Node        DB       Apache
    -----ws--->
                   -----ws--->
                                   ----SQL--->
                   <--302----
                   ----------------------------ws-------------->
   <-------------------------101-----------------------------
```

This is, a Node WS server to do async lookup and reply Nginx with 302 with the exact WS server it must forward the WS connection.

**Comment by alexhultman** [comment]
Yes I understand but it is very low priority. Esp so if you have a proxy which can do the auth itself

**Comment by ibc** [comment]
May I know which the problem of having ws.accept() and ws.reject() is? Node-WebSocket has them.

**Comment by ibc** [comment]
BTW my use case above is not about auth. It's about routing the WS connection to a dynamically obtained destination.

**Comment by phablulo** [comment]
I have the same use case as @ibc
